Add UI E2E test for DAGs list display

chen0427ok opened the pull request:
## Description

Implements UI E2E test case **DAGS-001** to verify DAGs list displays correctly, as requested in issue #59307.

## Changes

- **New test spec**: `dags-list.spec.ts` with 3 comprehensive test cases
- **Extended DagsPage**: Added methods to verify DAG list display functionality
- **Best practices**: Used `data-testid` selectors as recommended in the [E2E Testing README](https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/ui/tests/e2e/README.md#contributing-new-tests)
- **Improved stability**: Increased LoginPage timeouts for better reliability during test execution

## Test Cases

1. **Verify DAGs list displays after login** - Ensures the DAG list is visible and contains at least one DAG
2. **Verify DAG links are formatted correctly** - Validates all DAG links follow the expected pattern `/dags/{dag_id}`
3. **Verify specific test DAG exists in the list** - Confirms the test DAG (`example_bash_operator`) is present

## Test Results

**Local Testing Results:**

- **Chromium**: 3/3 tests passed
- **WebKit**: 3/3 tests passed
- **Firefox**: 0/3 tests failed

### Firefox Issue Details

All Firefox tests fail with the same error:
```
TimeoutError: locator.waitFor: Test timeout of 120000ms exceeded.
Call log:
  - waiting for locator('input[name="username"]') to be visible
```

**Root Cause**: Firefox browser receives a `404 Not Found` nginx error page when navigating to `/auth/login`, while Chromium and WebKit successfully load the login page. This appears to be an environment/configuration issue rather than a test code problem, as evidenced by:

1. Screenshot shows nginx 404 error page
2. Same test code works perfectly in Chromium and WebKit
3. Issue occurs consistently across all three Firefox test cases

**Screenshots**: Test results and Firefox error screenshots are available in `test-results/` directory.
<img width="1920" height="1080" alt="test-failed-1" src="https://github.com/user-attachments/assets/0f508c0f-c184-4b71-96c5-b3b592a15313" />

## Follows Contribution Guidelines

- Uses Page Object Model (POM) pattern
- Uses `data-testid` selectors where available
- Test names start with "should"
- Tests are independent and set up their own state
- Includes meaningful assertions beyond navigation checks
- Follows existing code style and conventions

## Related Issue

Closes #59307

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary
- [x] I have made corresponding changes to the documentation (README.md already includes guidance)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/LoginPage.ts:
```diff
@@ -84,6 +84,9 @@ export class LoginPage extends BasePage {
     await this.navigateTo(LoginPage.loginUrl);
+
+    // Wait for login form to be fully loaded (especially important for first test run)
+    await this.usernameInput.waitFor({ state: "visible", timeout: 120_000 });
```
I think you can try the command below:
```
breeze testing ui-e2e-tests --test-pattern "dags-list.spec.ts"
```
If there are no issues there, we can improve this part. :)

chen0427ok replied:
@choo121600 Thanks for the suggestion! I've run the tests locally with `breeze testing ui-e2e-tests --test-pattern "dags-list.spec.ts"` and all tests are now passing:

  **Test Results:**
  - Chromium: 3/3 passed
  - Firefox: 3/3 passed (previously failed, now working!)
  - WebKit: 3/3 passed

  **Total: 9 passed in 54.8s**

  The increased timeouts in `LoginPage.ts` (60s for navigation, 120s for form load) successfully resolved the Firefox issues. The tests are ready for review!

choo121600 replied:
I don’t think we need such a long timeout. It was enugh before.

pierrejeambrun commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -0,0 +1,108 @@
+  });
+
+  test("should display DAGs list after successful login", async () => {
+    test.setTimeout(2 * 60 * 1000); // 2 minutes timeout
```
Is there a way to set this for all 3 tests here. (maybe in the before each?)
```
test.beforeEach(async ({ page }, testInfo) => {
  // Extend timeout for all tests running this hook by 30 seconds.
  testInfo.setTimeout(2 * 60 * 1000);
});
```

pierrejeambrun reviewed (approved):
Thanks for the PR 🎉 

One nit.

LGTM

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -98,3 +99,289 @@ test.describe("Dag Trigger Workflow", () => {
+    await dagsPage.switchToCardView();
+
+    // Step 5: Verify card view is displayed
+    const cardViewVisible = await dagsPage.verifyCardViewVisible();
```
Can we also check if DAGs are also visible in card view?

chen0427ok replied:
Great suggestion! I've added DAG visibility verification in both views.                                                                                                                                           
                                                                                                                                                                                                                        
## Changes Made                                                                                                                                                                                                   
                                                                                                                                                                                                                  
Now the test verifies:                                                                                                                                                                                            
1. Card view container is displayed                                                                                                                                                                            
2. **DAGs are actually visible in card view** (new)                                                                                                                                                            
3. Table view container is displayed                                                                                                                                                                           
4. **DAGs are actually visible in table view** (new)                                                                                                                                                           
                                                                                                                                                                                                                        
The test now uses \`getDagsCount()\` to ensure \`> 0\` DAGs are rendered in each view mode, confirming that the view toggle not only changes the container but also properly displays the DAGs.                   
                                                                                                                                                                                                                        Thanks for the feedback!

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/LoginPage.ts:
```diff
@@ -67,7 +67,7 @@ export class LoginPage extends BasePage {
           return !urlString.includes("/login");
         },
-        { timeout: 15_000 },
+        { timeout: 60_000 },
```
Do we really need this much timeout?

vatsrahul1001 replied:
Yeah 15_000 is sufficient

vatsrahul1001 replied:
@chen0427ok can you fix this

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -98,3 +99,299 @@ test.describe("Dag Trigger Workflow", () => {
 });
+
+/**
+ * DAGS-001: Verify Dags List Displays
```
```suggestion
 * Verify Dags List Displays
```

choo121600 replied:
It seems we don’t need to include the issue task number as a comment.

choo121600 reviewed (commented):
There seem to be many unnecessary timeouts. I’m not an expert either, so could you explain why they were added?

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/LoginPage.ts:
```diff
@@ -67,7 +67,7 @@ export class LoginPage extends BasePage {
           return !urlString.includes("/login");
         },
-        { timeout: 15_000 },
+        { timeout: 60_000 },
```
```suggestion
        { timeout: 15_000 },
```

vatsrahul1001 replied:
@chen0427ok unresolved this comment as this is still not fixed

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -75,16 +122,133 @@ export class DagsPage extends BasePage {
+    // Select all DAG links using href pattern
+    const dagLinks = this.page.locator('a[href^="/dags/"]');
+
+    // Wait for page to load - check for either DAG links or "No Dag found" message
```
Please change the parts labeled “DAG” to “Dag.”
We’ve agreed to use “Dag” instead of “DAG” going forward :)
Also, please update any other occurrences to use “Dag” as well.

chen0427ok replied:
Thanks, I've updated all occurrences to use "Dag".

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -61,8 +61,6 @@ test.describe("Dag Trigger Workflow", () => {
   });
 
   test("should successfully trigger a Dag run", async () => {
-    test.setTimeout(7 * 60 * 1000);
```
The Dag Trigger Workflow test is flaky in CI - it times out at 30 seconds before the DAG run completes.
The original test had test.setTimeout(7 * 60 * 1000) (7 minutes) to allow enough time for the DAG to run. This line was removed in this PR, causing the test to fall back to the default 30-second timeout.
Please restore the timeout

vatsrahul1001 replied:
See flakines here https://github.com/apache/airflow/actions/runs/20450887934/job/58764983378?pr=59374

chen0427ok replied:
Thanks for pointing this out!
I’ve restored test.setTimeout(7 * 60 * 1000); in the “should successfully trigger a Dag run” test so it has enough time to complete and avoids the 30-second default timeout causing flakiness.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +83,164 @@ test.describe("Dag Details Tab", () => {
+
+    const failedCount = await dagsPage.getDagsCount();
+
+    expect(failedCount).toBeGreaterThanOrEqual(0);
```
This uses toBeGreaterThanOrEqual(0) which will pass even when filtering returns zero results. This doesn't actually verify that filtering works correctly. 
The test should assert toBeGreaterThan(0) to confirm filtering actually returns matching DAGs. However, this requires pre-requisite steps to ensure DAGs exist in different states:
For success and failed filters: Trigger a DAG run and wait for completion before testing these filters. This ensures we have at least one DAG in each state.
For running and queued filters: These are trickier to test reliably since the states are transient(Could be a follow-up improvement.)

vatsrahul1001 replied:
@chen0427ok Is this fixed?

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +83,164 @@ test.describe("Dag Details Tab", () => {
+
+    await dagsPage.clickSortSelect();
+
+    await expect(dagsPage.page.getByRole("option").first()).toBeVisible();
```
@chen0427ok Test verifies the sort dropdown is visible and opens, but doesn't verify that sorting actually works. Consider expanding the test to verify sorting behavior, maybe something like below
```
  test("should sort Dags by name in card view", async () => {
    await dagsPage.navigate();
    await dagsPage.verifyDagsListVisible();

    await dagsPage.switchToCardView();

    await expect(dagsPage.sortSelect).toBeVisible();

    const ascNames = await dagsPage.getDagNames();

    expect(ascNames.length).toBeGreaterThan(1);

    await dagsPage.clickSortSelect();

    await expect(dagsPage.page.getByRole("option").first()).toBeVisible();

    await dagsPage.page.getByRole("option", { name: "Sort by Display Name (Z-A)" }).click();

    await dagsPage.page.waitForTimeout(500);

    const descNames = await dagsPage.getDagNames();

    // Verify order changed
    expect(descNames[0]).not.toEqual(ascNames[0]);

    // Verify Z-A direction: first item should come after last item alphabetically
    expect(descNames[0].localeCompare(descNames[descNames.length - 1])).toBeGreaterThan(0);
  });
```

chen0427ok replied:
Thanks a lot for the detailed example! I’ll extend the existing sort test to verify the ordering logic (using an approach similar to what you suggested) and push the updates to this PR.

vatsrahul1001 replied:
Thanks @chen0427ok we can merge post it

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +83,179 @@ test.describe("Dag Details Tab", () => {
+    await dagsPage.verifyDagsListVisible();
+    const finalCount = await dagsPage.getDagsCount();
+
+    expect(finalCount).toBe(initialCount);
```
It looks like the E2E tests are failing here. Could you please take a look?

chen0427ok replied:
It seems like the count of DAGs did not revert to the original count. Likely, clearing the search isn't resetting the filter/UI state fully. I would find the solution to the problem then push an update to address the issues.

chen0427ok committed:
Fix race conditions in pagination for slower browsers

- Add API interception before pagination clicks
- Increase timeout from 10s to 30s for Firefox compatibility
- Add test.setTimeout(120_000) to pagination and search tests
- Improve waitForDagList timeout to 30s

Add comprehensive E2E tests for DAGs list page:
- Dags List Display: verify list visibility, dag links, dag existence
- Dags View Toggle: switch between card and table views
- Dags Search: search and clear with race condition handling
- Dags Status Filtering: filter by success/failed/running/queued
- Dags Sorting: sort by name in card view

chen0427ok committed:
fix(tests): Increase timeout for dags-list spec

Increases the timeout for the dags-list.spec.ts e2e test to prevent flakiness.

chen0427ok committed:
fix(tests): Update E2E test selectors to match current UI

chen0427ok committed:
fix(tests): Use expect.poll for flaky Dags Search test                                                Replace direct assertion with polling to handle UI rendering timing                                                  differences between local and CI environments.

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -333,8 +608,30 @@ export class DagsPage extends BasePage {
+    const cardList = this.page.locator('[data-testid="card-list"]');
+    const tableList = this.page.locator('[data-testid="table-list"]');
+
+    await expect(cardList.or(tableList)).toBeVisible({ timeout: 30_000 });
```
It looks like the CI failure is happening around this part.
Could you take a look here?

You can also run the following command locally to reproduce with the UI e2e tests:
```
breeze testing ui-e2e-tests --test-pattern "dags-list.spec.ts"
```
This might help with debugging

chen0427ok replied:
"Thanks for the guidance! I realized the CI failure was caused by the test timing out when the filter result was empty (displaying 'No Dag found' instead of a list).

I've updated waitForDagList in DagsPage.ts to correctly handle the 'No Dag found' state. I verified the fix locally using the command you provided (breeze testing ui-e2e-tests --test-pattern "dags-list.spec.ts"), and all tests passed successfully."
<img width="1218" height="842" alt="截圖 2026-01-26 下午2 55 56" src="https://github.com/user-attachments/assets/53f916f3-4b25-465d-bd04-788db7af2122" />
<img width="1230" height="852" alt="截圖 2026-01-26 下午2 56 07" src="https://github.com/user-attachments/assets/48706c25-c0c1-4fef-b4e2-e7cdf7286442" />

choo121600 replied:
@vatsrahul1001 
It might be helpful to include this guide in the issue description so others can more easily reproduce and debug the problem. What do you think?

> You can also run the following command locally to reproduce with the UI e2e tests:
> 
> ```
> breeze testing ui-e2e-tests --test-pattern "dags-list.spec.ts"
> ```
> 
> This might help with debugging

vatsrahul1001 replied:
@chen0427ok, this is already updated in [readme](https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/ui/tests/e2e/README.md) here but yes we can add this is issue description as well

vatsrahul1001 replied:
@chen0427ok I still see tests are failing in CI though

chen0427ok committed:
fix(tests): Handle empty state in waitForDagList to prevent timeout

vatsrahul1001 committed:
Merge branch 'main' into test/ui-e2e-dags-list-display

choo121600 committed:
Merge branch 'main' into test/ui-e2e-dags-list-display

choo121600 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -331,8 +606,36 @@ export class DagsPage extends BasePage {
+    // Wait for either card-list, table-list, or empty state to be visible
+    const cardList = this.page.locator('[data-testid="card-list"]');
+    const tableList = this.page.locator('[data-testid="table-list"]');
+    const noDagFound = this.page.locator("text=No Dag found");
```
The test fails because the locator uses singular form "No Dag found" but the actual UI renders "No Dags found"

chen0427ok replied:
"Thanks for the feedback!

I've resolved the merge conflicts and updated the tests to be more robust:

Flexible Text Matching: I updated the locator to use a Regex (text=/no dags? found/i). This handles both singular ("Dag") and plural ("Dags") forms, as well as case sensitivity, preventing future flakiness if the UI text changes.

Trigger Button Fix: I switched from .nth(1) to .last() for the confirmation button. This ensures the test correctly targets the modal button even when the main page trigger button is icon-only (has no text).

I verified locally that all tests in dags-list.spec.ts are passing now. Ready for another look!"

chen0427ok committed:
fix(tests): Harden E2E locators and replace fixed delays with polling

- Use case-insensitive regex (/no dag/i) for empty state detection
  to handle singular/plural/casing variations
- Add fallback <table> locator in waitForDagList when data-testid
  attributes are missing
- Change confirmButton from .nth(1) to .last() to handle icon-only
  trigger buttons
- Replace waitForTimeout(500) with expect.poll in Dags Sorting test

choo121600 committed:
Merge branch 'main' into test/ui-e2e-dags-list-display

choo121600 committed:
Merge branch 'main' into test/ui-e2e-dags-list-display

choo121600 reviewed (approved):
All CI checks have passed, and the tests look good.
Thanks for your hard work!
